src/com/github/Kraken3/AFKPGC/BotDetector.java

```diff
@@ -0,0 +1,306 @@
+								/* Because the ban time is based on real time and the next run
+								 * of this method is based on the tick, the ban time needs to be
+								 * adjusted to the tick. The long form of this would be:
+								 * (20/currentTPS) * (frequency/20) * 1000
+								 * The time is in ms */
+								BanEntry leBan = banList.addBan(curSuspect.getName(),
+										"You were suspected to cause lag and banned for "
+												+ (bantime / 1000) + " seconds",
+												new Date(currentDate.getTime() + bantime), null);
+								Player p = Bukkit.getPlayer(curSuspect.getUUID());
+								if (p != null) {
+					   				p.kickPlayer(leBan.getReason());
+								}
+								// ban the player briefly and skip the other suspects.
+								AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+										") exceeded ban threshold with ", ls.getLagCompute(), " banned for",
+										bantime / 1000, " milliseconds");
+								break;
+							} else {
+								AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+										") cleared via insufficient lagsources [", ls.getLagCompute(), "]");
+							}
+						} else {
+							AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+									") cleared via bounding box [", truebot, "]");
+						}
+						// It passed the truebot(tm) detection -- for now. Give a temporary reprieve.
+						reprieve.put(curSuspect.getUUID(), maxReprieve);
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " unlikely bot, given reprieve (bounds test ",
+								truebot, "): ", bounds);
+					} else {
+						// else not enough info yet. Pass.
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " Not enough bounds data, skip for now");
+					}
+				}
+			} else {
+				AFKPGC.debug("No suspects this round.");
+			}
+
+			if (lastRoundSuspect != null) {
+				// TODO: This is starting to stink of synchronization issues. We shouldn't need
+				//       to check immune so often within the same compute round.
+				if (!AFKPGC.immuneAccounts.contains(lastRoundSuspect.getUUID())) {
+					if (currentTPS - lastRoundTPS > criticalTPSChange) {
+						/* This can be relatively sensitive, because it will only ban players for 
+						 * a longer period of time, if it catches them twice here */
+						Date currentDate = new Date();
+						if (suspectedBotters.contains(lastRoundSuspect.getName())) {
+							if (longBans) {
+								BanEntry leBan = banList.addBan( lastRoundSuspect.getName(),
+										"Kicking you resulted in a noticeable TPS improvement, so you " +
+										"were banned until the TPS goes back to normal values.",
+										new Date(currentDate.getTime() + longBan),
+										null); // long ban.
+								Player p = Bukkit.getPlayer(lastRoundSuspect.getUUID());
+								if (p != null) {
+					   				p.kickPlayer(leBan.getReason());
+								}
+								bannedPlayers.add(lastRoundSuspect.getName());
+								addToBanfile(lastRoundSuspect.getName());

```

Comment:
It ... is in the if clause? I'm not sure what you're referring to here, if longbans is off no-one gets added to the ban file.